crypto/hacl_hash.h

```diff
@@ -0,0 +1,392 @@
+	(uint64_t)0x391c0cb3c5c95a63U, (uint64_t)0x4ed8aa4ae3418acbU,
+	(uint64_t)0x5b9cca4f7763e373U, (uint64_t)0x682e6ff3d6b2b8a3U,
+	(uint64_t)0x748f82ee5defb2fcU, (uint64_t)0x78a5636f43172f60U,
+	(uint64_t)0x84c87814a1f0ab72U, (uint64_t)0x8cc702081a6439ecU,
+	(uint64_t)0x90befffa23631e28U, (uint64_t)0xa4506cebde82bde9U,
+	(uint64_t)0xbef9a3f7b2c67915U, (uint64_t)0xc67178f2e372532bU,
+	(uint64_t)0xca273eceea26619cU, (uint64_t)0xd186b8c721c0c207U,
+	(uint64_t)0xeada7dd6cde0eb1eU, (uint64_t)0xf57d4f7fee6ed178U,
+	(uint64_t)0x06f067aa72176fbaU, (uint64_t)0x0a637dc5a2c898a6U,
+	(uint64_t)0x113f9804bef90daeU, (uint64_t)0x1b710b35131c471bU,
+	(uint64_t)0x28db77f523047d84U, (uint64_t)0x32caab7b40c72493U,
+	(uint64_t)0x3c9ebe0a15c9bebcU, (uint64_t)0x431d67c49c100d4cU,
+	(uint64_t)0x4cc5d4becb3e42b6U, (uint64_t)0x597f299cfc657e2aU,
+	(uint64_t)0x5fcb6fab3ad6faecU, (uint64_t)0x6c44198c4a475817U
+};
+
+void Hacl_SHA2_Scalar32_sha256_init(uint32_t *hash);
+
+void Hacl_SHA2_Scalar32_sha256_update_nblocks(uint32_t len, uint8_t *b,
+					      uint32_t *st);
+
+void Hacl_SHA2_Scalar32_sha256_update_last(uint64_t totlen, uint32_t len,
+					   uint8_t *b, uint32_t *hash);
+
+void Hacl_SHA2_Scalar32_sha256_finish(uint32_t *st, uint8_t *h);
+
+void Hacl_SHA2_Scalar32_sha224_init(uint32_t *hash);
+
+void Hacl_SHA2_Scalar32_sha224_update_last(uint64_t totlen, uint32_t len,
+					   uint8_t *b, uint32_t *st);
+
+void Hacl_SHA2_Scalar32_sha224_finish(uint32_t *st, uint8_t *h);
+
+void Hacl_SHA2_Scalar32_sha512_init(uint64_t *hash);
+
+void Hacl_SHA2_Scalar32_sha512_update_nblocks(uint32_t len, uint8_t *b,
+					      uint64_t *st);
+
+void Hacl_SHA2_Scalar32_sha512_update_last(FStar_UInt128_uint128 totlen,
+					   uint32_t len, uint8_t *b,
+					   uint64_t *hash);
+
+void Hacl_SHA2_Scalar32_sha512_finish(uint64_t *st, uint8_t *h);
+
+void Hacl_SHA2_Scalar32_sha384_init(uint64_t *hash);
+
+void Hacl_SHA2_Scalar32_sha384_update_nblocks(uint32_t len, uint8_t *b,
+					      uint64_t *st);
+
+void Hacl_SHA2_Scalar32_sha384_update_last(FStar_UInt128_uint128 totlen,
+					   uint32_t len, uint8_t *b,
+					   uint64_t *st);
+
+void Hacl_SHA2_Scalar32_sha384_finish(uint64_t *st, uint8_t *h);
+
+typedef Hacl_Streaming_MD_state_32 Hacl_Streaming_SHA2_state_sha2_224;
+
+typedef Hacl_Streaming_MD_state_32 Hacl_Streaming_SHA2_state_sha2_256;
+
+typedef Hacl_Streaming_MD_state_64 Hacl_Streaming_SHA2_state_sha2_384;
+
+typedef Hacl_Streaming_MD_state_64 Hacl_Streaming_SHA2_state_sha2_512;
+
+/**
+Allocate initial state for the SHA2_256 hash. The state is to be freed by
+calling `free_256`.
+*/
+Hacl_Streaming_MD_state_32 *Hacl_Streaming_SHA2_create_in_256(void);
+
+/**
+Copies the state passed as argument into a newly allocated state (deep copy).
+The state is to be freed by calling `free_256`. Cloning the state this way is
+useful, for instance, if your control-flow diverges and you need to feed
+more (different) data into the hash in each branch.
+*/
```

Comment:
About the format of comments


```suggestion
/*
 * Copies the state passed as argument into a newly allocated state (deep copy).
 * The state is to be freed by calling `free_256`. Cloning the state this way is
 * useful, for instance, if your control-flow diverges and you need to feed
 * more (different) data into the hash in each branch.
 */
```